### DISCUSSION TITLE
Is it possible to Implement monad memo as an effect? more generally can we branch on existing effects?

### DISCUSSION THREAD

**Discussion by eldritch-cookie** [haskell-effectful/effectful#212] [Q&A]
[monad-memo](https://hackage.haskell.org/package/monad-memo) is a library to memoize computations in monad transformers stacks,
The interesting thing is that each instance depends on the underlying transformer, so an effect that correctly implements monad memo would need to be "aware" of all other effects,
So i wanted to make something more simple first can we get an effect e that returns an integer saying how many effects there are?
this seems easy given that in a internal module there is a sizeEnv function.
if so can we store a value in each effect layer?
i am fairly certain i could hack something with typeclasses but then we would fall in the N^2 instances problem wouldn't we? also i am fairly certain that such a solution would have to deal with defunctionalization symbols due to needing to apply a type family to a list of effects.

**Comment by arybczak** [comment]
> The interesting thing is that each instance depends on the underlying transformer

Where is this implemented? I see `MonadMemo` has the usual trivial pass-through instances for standard transformers,  didn't find anything else.

**Reply by eldritch-cookie** [reply]
the thing is the instances aren't trivial pass-through, each monad transformer adds more to the info memoized, for example the StateT instance memoizes a tuple of s and k instead of only k.

**Reply by arybczak** [reply]
Oh right, I didn't read it properly :( Maybe it could be done for `Eff` as a low-level primitive effect (you probably would need to memorize all effect data types from the `Storage` for it to work reliably), but it looks highly non-trivial and I can't offer any help at the moment since I don't even know how the original library works, sorry.
